en/peripherals/companion_computer_peripherals.md

```diff
@@ -44,10 +44,23 @@ Cameras are among the most common form of data collection on a drone whether it
 
 ### Stereo Cameras
 
-Stereo cameras are typically used for depth perception, path planning and SLAM. They are in no way guaranteed to be plug and play with your companion computer.
+Stereo cameras are typically used for depth perception, path planning and SLAM. 
+They are in no way guaranteed to be plug and play with your companion computer.
 
 #### Where To Buy
 * [Intel® RealSense™ Depth Camera D435](https://click.intel.com/intelr-realsensetm-depth-camera-d435.html)
 * [Intel® RealSense™ Depth Camera D415](https://click.intel.com/intelr-realsensetm-depth-camera-d415.html)
 * [DUO MLX ](https://duo3d.com/product/duo-minilx-lv1)
 * [Ironsides](https://www.perceptin.io/products)
+
+
+## Data Telephony (LTE) {#data_telephony}
+
+An LTE USB module can be attached to a companion computer and used to route MAVLink traffic between the flight controller and the Internet.
+
+There is no standard approach for a ground station and companion to connect (i.e. for the GCS to discover the vehicle IP address, and visa versa).
```

Comment:
@bkueng I know this section is "weak" because I don't fully understand the approach discussed in https://github.com/PX4/px4_user_guide/issues/439#issuecomment-465037494

As I understand it there are two ways you can connect to a ground station:
1. UDP - QGC listens for UDP traffic on its IP address at port 14550. 
   - So if the vehicle knows the ground station IP address it can output UDP to that address and port. Is that correct?
   - Presumably you would have to change the code in PX4 to specify this address?
2. TCP- the GCS can connect via TCP to a known IP address/port. However AFAIK PX4 doesn't support TCPIP for this connection - so this wouldn't work?

And you talked about VPN. So as I understand you are suggesting you would set up VPN on the vehicle and connect to some VPN server "somewhere". Then presumably you still have to do "1" above to connect to the GCS?

Sorry if I'm being dumb.

Reply:
>     * So if the vehicle knows the ground station IP address it can output UDP to that address and port. Is that correct?

Correct.

>     * Presumably you would have to change the code in PX4 to specify this address? Or could you use MAVLink router?

Yes you have to use some kind of routing on the companion, mavlink-router is the canonical approach.

>     * Doesn't this imply you have to know the address of the GCS or service at configuration time?

Correct, for that approach. This is another reason why a VPN (with fixed IP's) helps you. If you use a VPN you can also use broadcasting, then the GCS IP does not need to be known.

>     2\. TCP- the GCS can connect via TCP to a known IP address/port. However AFAIK PX4 doesn't support TCPIP for this connection - so this wouldn't work?

It does (also mavlink-router), but it won't auto-connect AFAIK. So we generally use UDP.

> And you talked about VPN. So as I understand you are suggesting you would set up VPN on the vehicle and connect to some VPN server "somewhere". Then presumably you still have to do "1" above to connect to the GCS?

There does not need to be a VPN server, depending on the VPN. But for sure you will have to setup the VPN on the vehicle and the GCS machine. The VPN does 2 things:
- provide a fixed IP addresses for the GCS
- provide visibility: allows traffic to be sent directly to a GCS. If you want to route directly through the internet instead, and even if you know/have a fixed IP for the GCS, you cannot simply send MAVLink data to it, because of [NAT](https://en.wikipedia.org/wiki/Network_address_translation). If you want to do that you have to use a server in between (or some more advanced networking which we don't support).

The above approaches are for a direct vehicle-to-GCS connection, but it's possible to route MAVLink though a server if you want to (using mavlink-router on the server as well).

Reply:
Thanks @bkueng 

> You are touching a bigger topic here and I'm not sure how we should limit the scope to avoid having to document every possible option in detail.
Though I think it makes sense to document one or two options in more detail.

Indeed! More questions of course. 

I know only high level "consumer" stuff about VPNs - ie you set up a VPN client on your computer, it takes over your network connection and routes all your IP traffic encrypted to their server which then handles the request. To you it just looks like a normal internet connection. 
Given that, you'll understand my "how" questions below.

> There does not need to be a VPN server, depending on the VPN. But for sure you will have to setup the VPN on the vehicle and the GCS machine. The VPN does 2 things:
- provide a fixed IP addresses for the GCS
- provide visibility: allows traffic to be sent directly to a GCS. 

On the companion end, 
- I assume that the VPN automatically just routes all traffic - ie you don't have to do anything in mavlink router to get traffic to go via it?
- What I think you are saying is that:
  -if you do a UDP broadcast then this will be sent via the VPN tunnel and broadcast on the network running the GCS computer. That means every computer on the network including the GCS will get the message on port you set (presumably 14550 or whatever the default GCS port is). 
    - Is that right? 
    - Is it a problem that every computer on that same remote network also gets the UDP traffic, or is the "network" here restricted to just the remote computer running VPN/GCS?
  - You might otherwise send UDP to specific IP address of remote computer, but then intel router would need to know it in advance - correct? 
    - YOu say `provide fixed IP address for the GCS` - not quite sure how VPN helps here?

> If you want to route directly through the Internet instead, and even if you know/have a fixed IP for the GCS, you cannot simply send MAVLink data to it, because of [NAT](https://en.wikipedia.org/wiki/Network_address_translation). 

How does NAT hurt you? 

Anyway, once I see how VPN helps in big picture, would be good to know:
1. How to config intel router to forward traffic to UDP broadcast via VPN
2. What VPN solutions we have tried
3. What their setup looks like - ie what you need to know to connect the two ends.

Reply:
PS Perhaps we should also add this option to http://dev.px4.io/en/ros/offboard_control.html - removing the LTE part of course and just calling the channel "Internet".

Reply:
>     * I assume that the VPN automatically just routes all traffic - ie you don't have to do anything in mavlink router to get traffic to go via it?

Correct. You only need to configure to use the VPN IP of the GCS.

>     * Is that right?

No, it's broadcast within the VPN network. You can look at the VPN as a virtual local network. It bypasses the 'real' local network of a GCS. However a broadcast within the VPN will reach all clients connected to the VPN.

>     * or is the "network" here restricted to just the remote computer running VPN/GCS?

Exactly.

>     * You might otherwise send UDP to specific IP address of remote computer, but then intel router would need to know it in advance - correct?

Correct.

>     * YOu say `provide fixed IP address for the GCS` - not quite sure how VPN helps here?

Because your IP within the VPN usually stays fixed. It has to be fixed, because you use it in the mavlink router config.

>     1. How to config intel router to forward traffic to UDP broadcast via VPN

Just set it to the broadcast IP of the VPN. It's usually x.x.x.255.

> PS Perhaps we should also add this option to http://dev.px4.io/en/ros/offboard_control.html - removing the LTE part of course and just calling the channel "Internet".

I don't think so. You typically want to run ROS via local network, because you usually have large amounts of data that needs to sent in real-time (low-latency).